File: apps/api/app/models/project.py
import uuid
from datetime import datetime
from sqlalchemy import String, Integer, DateTime, ForeignKey, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.db import Base
from app.models.base import TimestampMixin


DOCUMENT_KINDS = [
    "synopsis", "outline", "characters", "world",
    "style", "principles", "summary", "notes",
]


class Project(Base, TimestampMixin):
    __tablename__ = "projects"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    genre: Mapped[str] = mapped_column(String(100), default="")
    author_note: Mapped[str] = mapped_column(String(2000), default="")
    default_pov: Mapped[str] = mapped_column(String(100), default="")
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime)

    documents: Mapped[list["ProjectDocument"]] = relationship(
        back_populates="project", cascade="all, delete-orphan"
    )
    chapters: Mapped[list["Chapter"]] = relationship(  # noqa: F821
        back_populates="project", cascade="all, delete-orphan"
    )


class ProjectDocument(Base):
    __tablename__ = "project_documents"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id: Mapped[str] = mapped_column(String(36), ForeignKey("projects.id"), nullable=False)
    kind: Mapped[str] = mapped_column(String(50), nullable=False)
    title: Mapped[str] = mapped_column(String(200), default="")
    content: Mapped[str] = mapped_column(String, default="")
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, server_default=func.now(), onupdate=func.now(), nullable=False
    )

    project: Mapped["Project"] = relationship(back_populates="documents")

    __table_args__ = (
        UniqueConstraint("project_id", "kind", name="uq_project_document_kind"),
    )

File: apps/web/src/types/project.ts
export interface ProjectListItem {
  id: string;
  name: string;
  genre: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface ProjectDocument {
  id: string;
  kind: DocumentKind;
  title: string;
  content: string;
  sort_order: number;
  updated_at: string;
}

export interface Project {
  id: string;
  name: string;
  genre: string;
  author_note: string;
  default_pov: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
  documents: ProjectDocument[];
}

export type DocumentKind =
  | "synopsis"
  | "outline"
  | "characters"
  | "world"
  | "style"
  | "principles"
  | "summary"
  | "notes";

export const DOCUMENT_KINDS: DocumentKind[] = [
  "synopsis",
  "outline",
  "characters",
  "world",
  "style",
  "principles",
  "summary",
  "notes",
];

export const DOCUMENT_LABELS: Record<DocumentKind, string> = {
  synopsis: "梗概",
  outline: "大纲",
  characters: "人物",
  world: "世界观",
  style: "风格",
  principles: "原则",
  summary: "摘要",
  notes: "笔记",
};

export interface ProjectCreate {
  name: string;
  genre?: string;
  author_note?: string;
  default_pov?: string;
}

export interface ProjectUpdate {
  name?: string;
  genre?: string;
  author_note?: string;
  default_pov?: string;
}

export interface DocumentUpdate {
  title?: string;
  content?: string;
}
